test/functional/wallet_hd.py

```diff
@@ -153,5 +153,74 @@ def run_test(self):
+        self.nodes[1].createwallet(wallet_name='origin', blank=True)
+        self.stop_node(1)
+        self.start_node(1, extra_args=['-keypool=3', '-wallet=origin'])
+        connect_nodes(self.nodes[0], 1)
+
+        # sethdseed restoring and seeing txs to addresses out of the keypool
+        origin_rpc = self.nodes[1].get_wallet_rpc('origin')
+        seed = self.nodes[0].dumpprivkey(self.nodes[0].getnewaddress())
+        origin_rpc.sethdseed(True, seed)
+
+        self.nodes[1].createwallet(wallet_name='restore', blank=True)
+        restore_rpc = self.nodes[1].get_wallet_rpc('restore')
+        restore_rpc.sethdseed(True, seed)
+        restore_rpc.sethdseed(True)
+
+        # Check persistence of inactive seed
+        restore_rpc.unloadwallet()
+        self.nodes[1].loadwallet('restore')
+        restore_rpc = self.nodes[1].get_wallet_rpc('restore')
+
+        # Empty keypool and get an address that is beyond the initial keypool
+        origin_rpc.getnewaddress()
+        origin_rpc.getnewaddress()
+        last_addr = origin_rpc.getnewaddress()
+        addr = origin_rpc.getnewaddress()
+
+        # Check that the restored seed does not have addr but does have last_addr
+        info = restore_rpc.getaddressinfo(last_addr)
+        assert_equal(info['ismine'], True)
+        info = restore_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], False)
+        info = origin_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], True)
+
+        txid = self.nodes[0].sendtoaddress(addr, 1)
+        origin_rpc.sendrawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+        self.nodes[0].generate(1)
+        origin_rpc.gettransaction(txid)
+        assert_raises_rpc_error(-5, 'Invalid or non-wallet transaction id', restore_rpc.gettransaction, txid)
+        out_of_kp_txid = txid
+
+        txid = self.nodes[0].sendtoaddress(last_addr, 1)
+        origin_rpc.sendrawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+        self.nodes[0].generate(1)
+        origin_rpc.gettransaction(txid)
+        restore_rpc.gettransaction(txid)
+        assert_raises_rpc_error(-5, 'Invalid or non-wallet transaction id', restore_rpc.gettransaction, out_of_kp_txid)
+
+        restore_rpc.rescanblockchain()
+        restore_rpc.gettransaction(out_of_kp_txid)
+        info = restore_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], True)
+
+        # Check again that 3 keys were derived.
+        # Empty keypool and get an address that is beyond the initial keypool
+        origin_rpc.getnewaddress()
+        origin_rpc.getnewaddress()
+        last_addr = origin_rpc.getnewaddress()
+        addr = origin_rpc.getnewaddress()
+
+        # Check that the restored seed does not have addr but does have last_addr
+        info = restore_rpc.getaddressinfo(last_addr)
+        assert_equal(info['ismine'], True)
+        info = restore_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], False)
+
```

Comment:
for info I appended these checks after the rescan so that all checks before the rescan are also repeated after it, and all still pass
```diff
         assert_equal(info['ismine'], False)
+        info = origin_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], True)
+
+        txid = self.nodes[0].sendtoaddress(addr, 1)
+        origin_rpc.sendrawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+        self.nodes[0].generate(1)
+        origin_rpc.gettransaction(txid)
+        assert_raises_rpc_error(-5, 'Invalid or non-wallet transaction id', restore_rpc.gettransaction, txid)
+        out_of_kp_txid = txid
+
+        txid = self.nodes[0].sendtoaddress(last_addr, 1)
+        origin_rpc.sendrawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+        self.nodes[0].generate(1)
+        origin_rpc.gettransaction(txid)
+        restore_rpc.gettransaction(txid)
+        assert_raises_rpc_error(-5, 'Invalid or non-wallet transaction id', restore_rpc.gettransaction, out_of_kp_txid)
+
+        restore_rpc.rescanblockchain()
+        restore_rpc.gettransaction(out_of_kp_txid)
+        info = restore_rpc.getaddressinfo(addr)
+        assert_equal(info['ismine'], True)
 
 if __name__ == '__main__':
     WalletHDTest().main ()
```

Reply:
It was an informational comment only.